// cleverbot-node/lib/cleverbot.js
var crypto = require('crypto')
  , http = require('http')
  , Cleverbot = function(){
      this.params = Cleverbot.default_params;
    };

Cleverbot.default_params = {
  'stimulus'         : '' , 'start'      : 'y'    , 'sessionid' : '',
  'vText8'           : '' , 'vText7'     : ''     , 'vText6' : '',
  'vText5'           : '' , 'vText4'     : ''     , 'vText3' : '',
  'vText2'           : '' , 'icognoid'   : 'wsf'  , 'icognocheck' : '',
  'fno'              : '0', 'prevref'    : ''     , 'emotionaloutput' : '',
  'emotionalhistory' : '' , 'asbotname'  : ''     , 'ttsvoice' : '',
  'typing'           : '' , 'lineref'    : ''     , 'sub' : 'Say',
  'islearning'       : '1', 'cleanslate' : 'false',
};
Cleverbot.parserKeys = [
  'message', 'sessionid', 'logurl', 'vText8',
  'vText7', 'vText6', 'vText5', 'vText4',
  'vText3', 'vText2', 'prevref', '',
  'emotionalhistory', 'ttsLocMP3', 'ttsLocTXT', 'ttsLocTXT3',
  'ttsText', 'lineref', 'lineURL', 'linePOST',
  'lineChoices', 'lineChoicesAbbrev', 'typingData', 'divert'
];
Cleverbot.digest = function(body){
  var m = crypto.createHash('md5');
  m.update(body)
  return m.digest('hex');
};

Cleverbot.encodeParams = function(a1){
  var u=[];
  for(x in a1){
    if(a1[x] instanceof Array)
      u.push(x+"="+encodeURIComponent(a1[x].join(",")));
    else if(a1[x] instanceof Object)
      u.push(params(a1[x]));
    else
      u.push(x+"="+encodeURIComponent(a1[x]));
  }
  return u.join("&");
};



Cleverbot.prototype = {


  write2 : function(message,callback){

    proxyArr = ['177.69.67.245:8080', '183.230.127.59:8080', '183.221.41.152:8123', '86.122.124.11:80', '177.69.195.4:3128', '190.95.138.90:8080', '50.177.183.215:46371', '119.46.110.17:80', '91.185.215.141:9271', '199.241.137.180:8089', '91.185.215.141:9271', '119.46.110.17:80', '62.194.17.46:49082' , '87.98.184.7:3128', '186.94.60.66:8080', '24.165.13.134:28540', '91.185.215.141:60:72', '95.38.32.102:8080', '46.20.34.59:80','122.96.59.102:81', '213.42.212.209:8080', '91.185.215.141:7632', '218.28.96.39:3128', '86.97.125.17:8118', '190.38.89.107:8080', '190.207.186.75:8080', '111.161.126.84:80', '95.0.0.97:8080', '183.230.127.59:8085', '186.95.224.15:8080', '61.247.188.66:8080'];
 
     //proxyArr = ['190.38.89.107:8080', '190.207.186.75:8080', '111.161.126.84:80', '95.0.0.97:8080', '183.230.127.59:8085', '186.95.224.15:8080', '61.247.188.66:8080'];
 randNum =Math.floor((Math.random() * proxyArr.length) + 0);
proxyServ= "http://"+proxyArr[randNum];

    var clever = this;
    body = this.params;
    body.stimulus = message;
    body.icognocheck = Cleverbot.digest(Cleverbot.encodeParams(body).substring(9,35));
    console.log("using proxy: "+ proxyServ);
    var options = {
      
      host: 'www.personalityforge.com',
      port: 80,
      path: '/xmlchat.php?Message='+encodeURIComponent(message)+'&UserID=6&ResponseID=asking&Domain=www%2Emondobot%2Ecom',
      method: 'GET',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': Cleverbot.encodeParams(body).length
      }
    };
    responseHash = {};
    var req = http.request(options, function(res) {
      var cb = callback || function(){};
      res.on('data', function(chunk) {
        var chunk_data = chunk.toString();
          try{
          chunk_data=chunk_data.split("<response>")[1].split("</response>")[0];
            responseHash['message']= chunk_data;
        
          }
          catch(er){
            cb(undefined);
          }
        cb(responseHash);
      });
     });
    req.write(Cleverbot.encodeParams(body));
    req.end();
  },

  write : function(message,callback){

    proxyArr = ['177.69.67.245:8080', '183.230.127.59:8080', '183.221.41.152:8123', '86.122.124.11:80', '177.69.195.4:3128', '190.95.138.90:8080', '50.177.183.215:46371', '119.46.110.17:80', '91.185.215.141:9271', '199.241.137.180:8089', '91.185.215.141:9271', '119.46.110.17:80', '62.194.17.46:49082' , '87.98.184.7:3128', '186.94.60.66:8080', '24.165.13.134:28540', '91.185.215.141:60:72', '95.38.32.102:8080', '46.20.34.59:80','122.96.59.102:81', '213.42.212.209:8080', '91.185.215.141:7632', '218.28.96.39:3128', '86.97.125.17:8118', '190.38.89.107:8080', '190.207.186.75:8080', '111.161.126.84:80', '95.0.0.97:8080', '183.230.127.59:8085', '186.95.224.15:8080', '61.247.188.66:8080'];
 
     //proxyArr = ['190.38.89.107:8080', '190.207.186.75:8080', '111.161.126.84:80', '95.0.0.97:8080', '183.230.127.59:8085', '186.95.224.15:8080', '61.247.188.66:8080'];
 randNum =Math.floor((Math.random() * proxyArr.length) + 0);
proxyServ= "http://"+proxyArr[randNum];

    var clever = this;
    body = this.params;
    body.stimulus = message;
    body.icognocheck = Cleverbot.digest(Cleverbot.encodeParams(body).substring(9,35));
    console.log("using proxy: "+ proxyServ);
    var options = {
      'proxy':proxyServ,
      host: 'www.cleverbot.com',
      port: 80,
      path: '/webservicemin',
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': Cleverbot.encodeParams(body).length
      }
    };
    var req = http.request(options, function(res) {
      var cb = callback || function(){};
      res.on('data', function(chunk) {
        var chunk_data = chunk.toString().split("\r")
          , responseHash = {};
        for(var i = 0, iLen = chunk_data.length;i<iLen;i++){
          clever.params[Cleverbot.parserKeys[i]] = responseHash[Cleverbot.parserKeys[i]] = chunk_data[i];
        }
        cb(responseHash);
      });
     });
    req.write(Cleverbot.encodeParams(body));
    req.end();
  }
};

module.exports = Cleverbot;
